haas/api.py

```diff
@@ -726,6 +867,89 @@ def switch_delete_port(switch, port):
     db.session.delete(port)
     db.session.commit()
 
+
+@rest_call('GET', '/ports', Schema({}))
+def list_ports():
+    """List all ports.
+
+    Return a JSON of string contains ports.
+    """
+    get_auth_backend().require_admin()
+    ports = model.Port.query.all()
+    ports = sorted([p.label for p in ports])
+    return json.dumps(ports)
+
+
+@rest_call('GET', '/port/<path:port>', Schema({
+    'port': basestring,
+}))
+def show_port(port):
+    """Display nic and switch info related to this port.
+
+    If no nic or other network attachments, nic will be None.
+    """
+    port = _must_find(model.Port, port)
+
+    get_auth_backend().require_admin()
+    nic = model.Nic.query.filter(model.Nic.port_id == port.id).first()
+
+    result = {
+        'name': port.label,
+        'switch': port.owner.label,
+        'nic': nic.label,
+        #'niclist': niclist
+        #'node': node,
+        #'nics': nics
+    }
+
+    attachment = None
+    if nic is not None:
+        #nic = _must_find_n(node, model.Nic, nic)
+        #node = _must_find(model.Node, "sun-20")
+        #nics = _must_find(model.Nic, nic)
+        #result['nic'] = nic.label
+        result['node'] = nic.owner.label
+        for a in nic.attachments:
+            if nic.port.label == port.label:
+                attachment = a
+
+    if attachment is None:
+        result['attachment'] = None
+    else:
+        result['attachment'] = {'network': attachment.network.label,
+                                'channel': attachment.channel,
+        }
+
+    return json.dumps(result, sort_keys=True)
+
+@rest_call('POST', '/port/<path:port>/revert', Schema({
+    'port': basestring,
+}))
+def revert_port(port):
+    """Revert port to reinstate network attachments.
+
+    If there is no nic connected to the port, return.
+    """
+    get_auth_backend().require_admin()
+    port = _must_find(model.Port, port)
+    nic = model.Nic.query.filter(model.Nic.port_id == port.id).first()
+
+    if nic is None:
+        return '', 404
+
+    #delete old entry and create new oned
+    nic = _must_find(model.Nic, nic.label)
+    for a in nic.attachments:
+        db.session.query(model.NetworkAttachment)\
+                .filter_by(nic=a.nic, channel=a.channel)\
+                .delete()
+        db.session.add(model.NetworkingAction(nic=a.nic,
+                                      new_network=a.network,
+                                      channel=a.channel))
+        port.owner.revert(port)
```

Comment:
I'm thinking that the revert() has to happen before the for loop above, since we want the switch to have no VLANs attached, which the for loop above restores.